src/Extensions/ExtHostClient_Terminal.re

```diff
@@ -0,0 +1,101 @@
+open Oni_Core;
+
+module Log = (val Log.withNamespace("Oni2.Extensions.Terminal"));
+
+module ShellLaunchConfig = {
+  [@deriving show({with_path: false})]
+  type t = {
+    name: string,
+    executable: string,
+    arguments: list(string),
+  };
+
+  let to_yojson = ({name, executable, arguments}) => {
+    let args = arguments |> List.map(s => `String(s));
+    `Assoc([
+      ("name", `String(name)),
+      ("executable", `String(executable)),
+      ("arguments", `List(args)),
+    ]);
+  };
+};
+
+// MSG
+type msg =
+  | SendProcessTitle({
+      terminalId: int,
+      title: string,
+    })
+  | SendProcessData({
+      terminalId: int,
+      data: string,
+    })
+  | SendProcessPid({
+      terminalId: int,
+      pid: int,
+    })
+  | SendProcessExit({
+      terminalId: int,
+      exitCode: int,
+    });
+
+let handleMessage = (~dispatch, method, args) =>
+  switch (method) {
+  | "$sendProcessTitle" =>
+    switch (args) {
+    | [`Int(terminalId), `String(title)] =>
+      dispatch(SendProcessTitle({terminalId, title}))
+    | _ => Log.error("Unexpected arguments for $sendProcessTitle")
+    }
```

Comment:
I haven't gotten that far in the SCM project yet, but I've wanted to try using ocaml-decoders for this, since I think that'll remove the repetitive args pattern match and catch all error.

Reply:
Good to know! I was curious about this looking at the `ExtHostClient_SCM` module (which I modeled this after).

Reply:
I'll hold off on converting for now too... but it sounds like it would be a useful refactoring to reduce the repetitive error handling